ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Genesis/Governor.hs

```diff
@@ -93,4 +114,147 @@ updateLoEFragStall k getCandidates =
+          lowerBound = fromIntegral $ AF.length frag
+          unresolvedSlotsLB =
+            succWithOrigin $ AF.headSlot frag
+          unresolvedSlotsUB =
+              endOfGenesisWindow `min` claimedTip
+            where
+              claimedTip = succWithOrigin $ getTipSlotNo theirTip
+          hasBlockAfterGenesisWindow =
+            -- Note that if
+            -- > NotOrigin s = AF.headSlot candidateSuffix
+            -- this check is equivalent to
+            -- > s >= endOfGenesisWindow
+              succWithOrigin (AF.headSlot candidateSuffix)
+            > endOfGenesisWindow
+          upperBound =
+              lowerBound
+            + if hasBlockAfterGenesisWindow
+              then 0
+              else unSlotNo (intervalLength unresolvedSlotsLB unresolvedSlotsUB)
+          offersMoreThanK = AF.length candidateSuffix > fromIntegral k
+      pure (peer, (frag, offersMoreThanK, lowerBound, upperBound))
+
+    losingPeers = nubOrd $ do
+      (peer0 , (frag0, _        , _  , ub0)) <- Map.toList densityBounds
+      (_peer1, (frag1, moreThanK, lb1, _  )) <- Map.toList densityBounds
+      -- ensure that the two peer fragments don't share any
+      -- headers after the LoE
+      guard $ AF.lastPoint frag0 /= AF.lastPoint frag1
+      -- peer1 offers more than k blocks
+      guard moreThanK
+      -- peer1 definitely has higher density than peer0
+      guard $ lb1 > ub0
+      pure peer0
+
+    loeIntersectionSlot = AF.headSlot loeFrag
+    -- exclusive last slot in the Genesis window
+    endOfGenesisWindow =
+        succWithOrigin loeIntersectionSlot + SlotNo sgen
+
+    dropBeyondGenesisWindow =
+      AF.takeWhileOldest ((< endOfGenesisWindow) . blockSlot)
+
+    competingFrags =
+      Map.map dropBeyondGenesisWindow candidateSuffixes
+
+    -- Length of an interval with inclusive lower bound @a@ and exclusive upper
+    -- bound @b@.
+    intervalLength a b
+      | a <= b    = b - a
+      | otherwise = 0
+
+    showBounds :: (AnchoredFragment (Header blk), Bool, Word64, Word64) -> String
+    showBounds (_, more, lower, upper) = show lower ++ "/" ++ show upper ++ "[" ++ (if more then "+" else " ") ++ "]"
+
+{- TODO: convert this scribble into a useful explanatory diagram, illustrating the
+        density calculation below
+
+            |--------|
+
+    frag1: A - B - C - D - ...            <- exact
+            \
+    frag2:     E           claimed tip: E <- exact
+
+-}
+
+-- | Run the Genesis disconnection logic once.
+--
+--   * Maintain and update the LoE.
+--
+--   * Disconnect from peers with inferior density.
+updateLoEFragGenesis ::
+     forall m blk peer.
+     ( IOLike m
+     , Ord peer
+     , Show peer
+     , LedgerSupportsProtocol blk
+     )
+  => TopLevelConfig blk
+  -> Tracer m String
+  -> STM m (Map peer (AnchoredFragment (Header blk)))
+  -> STM m (Map peer (ChainSyncClientHandle m blk))
+  -> UpdateLoEFrag m blk
+updateLoEFragGenesis cfg tracer getCandidates getHandles =
```

Comment:
At the moment, the LoE fragment is updated every time a block is added to the chain db. 

An open question is how is LoE updated if all the blocks from all the peers have been received already.